src/jdk.httpserver/share/classes/sun/net/httpserver/ServerImpl.java

```diff
@@ -226,20 +226,36 @@ public final boolean isFinishing() {
         return finished;
     }
 
+    /**
+     * This method stops the server by adding a stop request event and
+     * waiting for the server until the event is triggered or until the maximum delay is triggered.
+     * <p>
+     * This ensures that the server is stopped immediately after all exchanges are complete,
+     * but still stops as soon as maximum delay is reached.
+     *
+     * @param delay maximum delay in stopping the server
+     */
     public void stop (int delay) {
         if (delay < 0) {
             throw new IllegalArgumentException ("negative delay parameter");
         }
+
+        logger.log(Level.TRACE, "stopping");
+        // posting a stop event, which will flip finished flag if it finishes
+        // before the timeout in this method
+        addEvent(new StopRequestedEvent());
+
         terminating = true;
         try { schan.close(); } catch (IOException e) {}
         selector.wakeup();
-        long latest = System.currentTimeMillis() + delay * 1000;
-        while (System.currentTimeMillis() < latest) {
+        long latest = System.nanoTime() + delay * 1000000000L;
+        while (System.nanoTime() < latest) {
```

Comment:
This is not correct as `latest` may overflow.

```suggestion
        long start = System.nanoTime();
        while ((System.nanoTime() - start) / 1000_000L < delay ) {
```

Reply:
Changed to the countdownlatch, so is removed in the next commit